hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java

```diff
@@ -0,0 +1,637 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.serialization.impl;
+
+import com.hazelcast.internal.serialization.impl.compact.DefaultCompactWriter;
+import com.hazelcast.nio.serialization.FieldType;
+import com.hazelcast.nio.serialization.GenericRecord;
+
+import java.util.Arrays;
+
+/**
+ * Should always be consistent with {@link com.hazelcast.nio.serialization.FieldType}
+ */
+public enum FieldOperations {
+
+    PORTABLE(new FieldTypeBasedOperations() {
+        @Override
+        public Object readObject(GenericRecord genericRecord, String fieldName) {
+            return ((InternalGenericRecord) genericRecord).getObject(fieldName);
+        }
+
+        @Override
+        public GenericRecord readInSerializedForm(GenericRecord genericRecord, String fieldName) {
+            return genericRecord.getGenericRecord(fieldName);
+        }
+
+        @Override
+        public void readFromGenericRecordToWriter(DefaultCompactWriter writer, GenericRecord genericRecord, String fieldName) {
+            writer.writeGenericRecord(fieldName, genericRecord.getGenericRecord(fieldName));
+        }
+    }),
+    BYTE(new FieldTypeBasedOperations() {
+        @Override
+        public Byte readObject(GenericRecord genericRecord, String fieldName) {
```

Comment:
Boxing occurs? Can we not eliminate this? if you separate the fixed and variable length field calls, would that help? You probably will need non-generic return, and use a big `if else` block.

Reply:
Boxing happens in the caller of these methods anyway. These methods are called in the query where we construct the Object and check equality. 
To eliminate this lots of query related classes should change. In my defense, this pr does not add new boxing. And this part, if needed can be optimized later.

Reply:
Since it is already being converted to Object in the caller method, then there is no point to optimize this but it needs to be optimized together with the Query system.